### PR TITLE
Add Recipe functionality

### LINKED ISSUE #4: As a user, I want to be able to create a recipe for an item so I can reference it in the future.
C for recipe class.

### PR DESCRIPTION
**Pull request by Rahul0506**
Refactor code base to be more general, add functionality to add recipes.
TODO:
- Delete Legacy
- Fix JavaDocs
- Fix and add test cases

### CODE REVIEW

File: src/main/java/seedu/address/InventoryMainApp.java [hunk 1]
```diff
@@ -90,38 +67,59 @@ public static Storage getLocationStorage() {
...
+            recipeListOptional = storage.readRecipeList();
+            if (recipeListOptional.isEmpty()) {
+                logger.info("Data file not found. Will be starting with a sample ItemList");
+            }
+            initialRecipeList = recipeListOptional.orElseGet(SampleDataUtil::getSampleRecipeList);
+        } catch (DataConversionException e) {
+            logger.warning("Data file not in the correct format. Will be starting with an empty ItemList");
+            initialRecipeList = new RecipeList();
+        } catch (IOException e) {
+            logger.warning("Problem while reading from the file. Will be starting with an empty ItemList");
```
Comment (aqcd) [inline src/main/java/seedu/address/InventoryMainApp.java]: ItemList->RecipeList

File: src/main/java/seedu/address/logic/Logic.java [hunk 2]
```diff
@@ -38,19 +40,15 @@
...
 
     ReadOnlyItemList getItemList();
 
     /** Returns an unmodifiable view of the filtered list of persons */
     ObservableList<Item> getFilteredItemList();
 
-    /**
-     * Returns the user prefs' address book file path.
-     */
+    /** Returns the user prefs' address book file path. */
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/Logic.java]: item list file path

File: src/main/java/seedu/address/logic/Logic.java [hunk 3]
```diff
@@ -59,6 +57,14 @@
 
     Path getLocationListFilePath();
 
+    ReadOnlyRecipeList getRecipeList();
+
+    /** Returns an unmodifiable view of the filtered list of persons */
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/Logic.java]: of recipes

File: src/main/java/seedu/address/logic/Logic.java [hunk 4]
```diff
@@ -59,6 +57,14 @@
...
+    ObservableList<Recipe> getFilteredRecipeList();
+
+    /** Returns the user prefs' address book file path. */
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/Logic.java]: recipe list file path

File: src/main/java/seedu/address/logic/LogicManager.java [hunk 5]
```diff
@@ -10,15 +10,13 @@
...
 import seedu.address.logic.commands.CommandResult;
 import seedu.address.logic.commands.exceptions.CommandException;
-import seedu.address.logic.parser.AddressBookParser;
+import seedu.address.logic.parser.InventoryParser;
 import seedu.address.logic.parser.exceptions.ParseException;
-import seedu.address.model.Model;
-import seedu.address.model.ReadOnlyAddressBook;
-import seedu.address.model.ReadOnlyItemList;
-import seedu.address.model.ReadOnlyLocationList;
+import seedu.address.model.*;
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/LogicManager.java]: will fail checkstyle

File: src/main/java/seedu/address/logic/parser/AddRecipeCommandParser.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+
+/**
+ * Parses input arguments and creates a new AddItemCommand object
+ */
+public class AddRecipeCommandParser implements Parser<AddRecipeCommand> {
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the AddItemCommand
+     * and returns an AddItemCommand object for execution.
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/parser/AddRecipeCommandParser.java]: item -> recipe

File: src/main/java/seedu/address/logic/parser/AddRecipeCommandParser.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import seedu.address.model.recipe.*;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static seedu.address.commons.core.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
+import static seedu.address.logic.parser.CliSyntax.*;
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/parser/AddRecipeCommandParser.java]: checkstyle

File: src/main/java/seedu/address/logic/parser/CliSyntax.java [hunk 8]
```diff
@@ -18,4 +18,10 @@
     public static final Prefix PREFIX_ITEM_DESCRIPTION = new Prefix("-d ");
     public static final Prefix PREFIX_ITEM_LOCATION = new Prefix("-l ");
 
+    // recipes start here
+    public static final Prefix PREFIX_RECIPE_PRODUCT_NAME = new Prefix("-n ");
+    public static final Prefix PREFIX_RECIPE_INGREDIENTS = new Prefix("-items ");
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/parser/CliSyntax.java:23]: can consider shorter prefix

**Review by aqcd — COMMENTED** [review]
Preliminary review. Mostly cosmetic since functionality isn't complete.

Initial apple having [1...6] for locationId and recipeId may not be intentional (bug?)
Initial recipes have item ids of items that don't exist.

File: src/main/java/seedu/address/logic/parser/CliSyntax.java (hunk 8 above)
Comment (Rahul0506) [reply src/main/java/seedu/address/logic/parser/CliSyntax.java:23]: This was specified in user guide, can be changed easily regardless

File: src/main/java/seedu/address/model/util/SampleDataUtil.java [hunk 9]
```diff
@@ -94,4 +108,24 @@ public static ReadOnlyLocationList getSampleLocationList() {
...
+
+    public static Recipe[] getSampleRecipes() {
+        List<Ingredient> ingredientList = new ArrayList<>();
+        ingredientList.add(new Ingredient(0, 2));
+        ingredientList.add(new Ingredient(1, 3));
+        IngredientList ingredients = new IngredientList();
+        ingredients.setItems(ingredientList);
+
+        return new Recipe[] {
+            new Recipe(0, ingredients, 2, new ProductQuantity("1"), "Recipe 1", false)
```
Comment (khenghun) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java]: Does not match constructor, should be 
new Recipe(0, ingredients, 2, "", new ProductQuantity("1"), "Recipe 1", false)
instead

File: src/main/java/seedu/address/model/recipe/Recipe.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, ingredients, productId, productQuantity, description, isDeleted);
+    }
+
+    @Override
+    public String toString() {
+        return String.format("Recipe for: %s. %s", productName, description); // <---------- TODO
```
Comment (khenghun) [inline src/main/java/seedu/address/model/recipe/Recipe.java]: Maybe update this or remove TODO?

File: src/main/java/seedu/address/model/recipe/ProductQuantity.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
+package seedu.address.model.recipe;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Recipe's product's quantity number in the inventory.
+ * Guarantees: immutable; is valid as declared in {@link #isValidQuantity(String)}
+ */
+public class ProductQuantity {
```
Comment (khenghun) [inline src/main/java/seedu/address/model/recipe/ProductQuantity.java]: can extend Quantity class if you want

**Review by khenghun — COMMENTED** [review]
just one bug, also need to remove test cases or update, LGTM

File: src/main/java/seedu/address/model/recipe/ProductQuantity.java (hunk 11 above)
Comment (Rahul0506) [reply src/main/java/seedu/address/model/recipe/ProductQuantity.java]: Good note, modified

**Pull request merged**